driver-core/src/main/com/mongodb/DBObjectCodec.java

```diff
@@ -211,43 +212,43 @@ private void writeValue(final BsonWriter bsonWriter, final EncoderContext encode
         if (value == null) {
             bsonWriter.writeNull();
         } else if (value instanceof DBRef) {
-            encodeDBRef(bsonWriter, (DBRef) value);
+            encodeDBRef(bsonWriter, (DBRef) value, encoderContext);
         } else if (value instanceof Map) {
-            encodeMap(bsonWriter, (Map<String, Object>) value);
+            encodeMap(bsonWriter, (Map<String, Object>) value, encoderContext);
         } else if (value instanceof Iterable) {
-            encodeIterable(bsonWriter, (Iterable) value);
+            encodeIterable(bsonWriter, (Iterable) value, encoderContext);
         } else if (value instanceof BSONObject) {
-            encodeBsonObject(bsonWriter, ((BSONObject) value));
+            encodeBsonObject(bsonWriter, (BSONObject) value, encoderContext);
         } else if (value instanceof CodeWScope) {
-            encodeCodeWScope(bsonWriter, (CodeWScope) value);
+            encodeCodeWScope(bsonWriter, (CodeWScope) value, encoderContext);
         } else if (value instanceof byte[]) {
             encodeByteArray(bsonWriter, (byte[]) value);
         } else if (value.getClass().isArray()) {
-            encodeArray(bsonWriter, value);
+            encodeArray(bsonWriter, value, encoderContext);
         } else if (value instanceof Symbol) {
             bsonWriter.writeSymbol(((Symbol) value).getSymbol());
         } else {
             Codec codec = codecRegistry.get(value.getClass());
-            codec.encode(bsonWriter, value, encoderContext);
+            encoderContext.encodeWithChildContext(codec, bsonWriter, value);
```

Comment:
Also make sure, here and below, to use the child context.  This has the effect of unsetting EncoderContext.encodingCollectibleDocument.  Decided it wasn't worth it to test all the paths to ensure that was happening, as it would require creating custom codecs for all of the types in order to assert.  Let me know if you disagree with this decision, and I can write some more tests.

Reply:
Ack